// cypress-ntlm-auth/dist/proxy/ntlm.proxy.mitm.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NtlmProxyMitm = void 0;
const net_1 = __importDefault(require("net"));
const url_converter_1 = require("../util/url.converter");
const inversify_1 = require("inversify");
const dependency_injection_types_1 = require("./dependency.injection.types");
const nodeCommon = require("_http_common");
let self;
let NtlmProxyMitm = class NtlmProxyMitm {
    constructor(configStore, portsConfigStore, connectionContextManager, winSsoFacade, negotiateManager, ntlmManager, upstreamProxyManager, debug) {
        this._configStore = configStore;
        this._portsConfigStore = portsConfigStore;
        this._connectionContextManager = connectionContextManager;
        this.WinSsoFacade = winSsoFacade;
        this._negotiateManager = negotiateManager;
        this._ntlmManager = ntlmManager;
        this._upstreamProxyManager = upstreamProxyManager;
        this._debug = debug;
        // Keep track of instance since methods will be triggered from HttpMitmProxy
        // events which means that 'this' is no longer the class instance
        self = this;
    }
    filterChromeStartup(ctx, errno, errorKind) {
        if (!ctx || !ctx.clientToProxyRequest || !errno) {
            return false;
        }
        const req = ctx.clientToProxyRequest;
        if (req.method === "HEAD" &&
            req.url === "/" &&
            req.headers.host &&
            req.headers.host.indexOf(".") === -1 &&
            (req.headers.host.indexOf(":") === -1 || req.headers.host.indexOf(":80") !== -1) &&
            req.headers.host.indexOf("/") === -1 &&
            errorKind === "PROXY_TO_SERVER_REQUEST_ERROR" &&
            errno === "ENOTFOUND") {
            self._debug.log("Chrome startup HEAD request detected (host: " + req.headers.host + "). Ignoring connection error.");
            return true;
        }
    }
    onError(ctx, error, errorKind) {
        var _a;
        if (self.filterChromeStartup(ctx, error.code, errorKind)) {
            return;
        }
        const url = ctx && ctx.clientToProxyRequest ? ctx.clientToProxyRequest.url : "";
        if (errorKind === "PROXY_TO_SERVER_REQUEST_ERROR") {
            // Act transparent - client will receive a network error instead of a 504
            self._debug.log(errorKind + " on " + url + ":", error, "Destroying client socket.");
            (_a = ctx.proxyToClientResponse.socket) === null || _a === void 0 ? void 0 : _a.destroy();
        }
        else {
            self._debug.log(errorKind + " on " + url + ":", error);
        }
    }
    isConfigApiRequest(targetHost) {
        if (!self._portsConfigStore.configApiUrl) {
            return false;
        }
        return targetHost.href.startsWith(self._portsConfigStore.configApiUrl);
    }
    onRequest(ctx, callback) {
        const targetHost = self.getTargetHost(ctx);
        if (targetHost) {
            let context = self._connectionContextManager.getConnectionContextFromClientSocket(ctx.clientToProxyRequest.socket);
            const useSso = self._configStore.useSso(targetHost);
            const useNtlm = useSso || self._configStore.exists(targetHost);
            if (context) {
                if (context.matchHostOrNew(targetHost) === false) {
                    self._debug.log("Existing client socket " +
                        context.clientAddress +
                        " received request to a different target, remove existing context");
                    self._connectionContextManager.removeAgent("reuse", context.clientAddress);
                    context = undefined;
                }
            }
            if (!context) {
                context = self._connectionContextManager.createConnectionContext(ctx.clientToProxyRequest.socket, ctx.isSSL, targetHost);
            }
            if (useNtlm) {
                self._debug.log("Request to " + targetHost.href + " in registered NTLM Hosts" + (useSso ? " (using SSO)" : ""));
                ctx.proxyToServerRequestOptions.agent = context.agent;
                context.clearRequestBody();
                ctx.onRequestData(function (ctx, chunk, callback) {
                    context.addToRequestBody(chunk);
                    return callback(undefined, chunk);
                });
            }
            else {
                if (self.isConfigApiRequest(targetHost)) {
                    self._debug.log("Request to config API");
                    ctx.proxyToServerRequestOptions.agent = self._connectionContextManager.getUntrackedAgent(targetHost);
                    context.configApiConnection = true;
                }
                else {
                    self._debug.log("Request to " + targetHost.href + " - pass on");
                    ctx.proxyToServerRequestOptions.agent = context.agent;
                }
            }
            return callback();
        }
        else {
            // The http-mitm-proxy cannot handle this scenario, if no target host header
            // is set it will get stuck in an infinite loop
            return callback(new Error('Invalid request - Could not read "host" header or "host" header refers to this proxy'));
        }
    }
    isNtlmProxyAddress(hostUrl) {
        if (!self._portsConfigStore.ntlmProxyPort) {
            return false;
        }
        return hostUrl.isLocalhost && hostUrl.port === self._portsConfigStore.ntlmProxyPort;
    }
    getTargetHost(ctx) {
        if (!ctx.clientToProxyRequest.headers.host) {
            self._debug.log('Invalid request - Could not read "host" header from incoming request to proxy');
            return null;
        }
        const host = ctx.clientToProxyRequest.headers.host;
        const hostUrl = url_converter_1.toCompleteUrl(host, ctx.isSSL, true);
        if (self.isNtlmProxyAddress(hostUrl)) {
            self._debug.log("Invalid request - host header refers to this proxy");
            return null;
        }
        return hostUrl;
    }
    getAuthMode(serverToProxyResponse, useSso) {
        if (serverToProxyResponse.statusCode !== 401 || !serverToProxyResponse.headers["www-authenticate"]) {
            return 0 /* NotApplicable */;
        }
        if (useSso && self._negotiateManager.acceptsNegotiateAuthentication(serverToProxyResponse)) {
            return 4 /* Negotiate */;
        }
        if (self._ntlmManager.acceptsNtlmAuthentication(serverToProxyResponse)) {
            return 3 /* NTLM */;
        }
        // TODO Basic auth
        return 1 /* NotSupported */;
    }
    onResponse(ctx, callback) {
        const targetHost = self.getTargetHost(ctx);
        if (!targetHost) {
            return callback();
        }
        const useSso = self._configStore.useSso(targetHost);
        const useNtlm = useSso || self._configStore.exists(targetHost);
        if (!useNtlm) {
            return callback();
        }
        const context = self._connectionContextManager.getConnectionContextFromClientSocket(ctx.clientToProxyRequest.socket);
        if (context && context.canStartAuthHandshake(targetHost)) {
            const authMode = self.getAuthMode(ctx.serverToProxyResponse, useSso);
            if (authMode === 0 /* NotApplicable */) {
                return callback();
            }
            if (authMode === 1 /* NotSupported */) {
                self._debug.log("Received 401 with unsupported protocol in www-authenticate header.", ctx.serverToProxyResponse.headers["www-authenticate"], "Ignoring.");
                return callback();
            }
            // Grab PeerCertificate for NTLM channel binding
            if (ctx.isSSL) {
                const tlsSocket = ctx.serverToProxyResponse.connection;
                const peerCert = tlsSocket.getPeerCertificate();
                // getPeerCertificate may return an empty object.
                // Validate that it has fingerprint256 attribute (added in Node 9.8.0)
                if (peerCert.fingerprint256) {
                    context.peerCert = peerCert;
                }
                else {
                    self._debug.log("Could not retrieve PeerCertificate for NTLM channel binding.");
                }
            }
            if (authMode === 4 /* Negotiate */) {
                self._debug.log("Received 401 with Negotiate in www-authenticate header. Starting handshake.");
                if (useSso) {
                    context.winSso = new self.WinSsoFacade("Negotiate", ctx.proxyToServerRequestOptions.host, context.peerCert);
                }
                self._negotiateManager.handshake(ctx, targetHost, context, (err, res) => self.handshakeCallback(ctx, err, res));
            }
            if (authMode === 3 /* NTLM */) {
                self._debug.log("Received 401 with NTLM in www-authenticate header. Starting handshake.");
                if (useSso) {
                    context.winSso = new self.WinSsoFacade("NTLM", ctx.proxyToServerRequestOptions.host, context.peerCert);
                }
                self._ntlmManager.handshake(ctx, targetHost, context, useSso, (err, res) => self.handshakeCallback(ctx, err, res));
            }
        }
        else {
            return callback();
        }
    }
    handshakeCallback(ctx, err, res) {
        var _a, _b;
        if (err) {
            self._debug.log("Cannot perform handshake.");
        }
        if (res) {
            if (ctx.clientToProxyRequest.headers["proxy-connection"]) {
                res.headers["proxy-connection"] = "keep-alive";
                if (res.statusCode && res.statusCode !== 401) {
                    res.headers["connection"] = "keep-alive";
                }
                else {
                    res.headers["connection"] = "close";
                }
            }
            ctx.proxyToClientResponse.writeHead(res.statusCode || 401, self.filterAndCanonizeHeaders(res.headers));
            res.on("data", (chunk) => ctx.proxyToClientResponse.write(chunk));
            res.on("end", () => ctx.proxyToClientResponse.end());
            res.resume();
        }
        else {
            // No response available, means that an network error occured on the connection to the target host
            // Signal this to generic error handling by destroying the original request socket
            (_a = ctx.proxyToServerRequest.socket) === null || _a === void 0 ? void 0 : _a.destroy();
            (_b = ctx.proxyToClientResponse.socket) === null || _b === void 0 ? void 0 : _b.destroy();
            /*
            const headers = ctx.serverToProxyResponse.headers;
            if (headers["proxy-connection"]) {
              headers["proxy-connection"] = "keep-alive";
              headers["connection"] = "close";
            }
            ctx.proxyToClientResponse.writeHead(401, self.filterAndCanonizeHeaders(headers));
            ctx.proxyToClientResponse.end();
            */
        }
    }
    onConnect(req, socket, head, callback) {
        if (!req.url) {
            self._debug.log("Invalid connect request - cannot read target url");
            return callback();
        }
        const targetHost = url_converter_1.toCompleteUrl(req.url, true, true);
        if (self._configStore.existsOrUseSso(targetHost)) {
            return callback();
        }
        if (self._upstreamProxyManager.hasHttpsUpstreamProxy(targetHost)) {
            // Don't tunnel if we need to go through an upstream proxy
            return callback();
        }
        // Let non-NTLM hosts tunnel through
        self._debug.log("Tunnel to", req.url);
        const onPrematureClose = function () {
            self._debug.log("cannot establish connection to server, CONNECT failed");
            socket.end("HTTP/1.1 502 Bad Gateway\r\n\r\n", "utf8");
        };
        const conn = net_1.default.connect({
            port: +targetHost.port,
            host: targetHost.hostname,
            allowHalfOpen: true,
        }, function () {
            conn.on("finish", () => {
                self._connectionContextManager.removeTunnel(socket);
                socket.destroy();
            });
            socket.on("close", () => {
                self._debug.log("client closed socket, closing tunnel to ", req.url);
                conn.end();
            });
            conn.removeListener("close", onPrematureClose);
            socket.write("HTTP/1.1 200 OK\r\n\r\n", "utf8", function () {
                conn.write(head);
                conn.pipe(socket);
                socket.pipe(conn);
                self._connectionContextManager.addTunnel(socket, conn);
            });
        });
        conn.once("close", onPrematureClose);
        conn.on("error", function (err) {
            filterSocketConnReset(err, "PROXY_TO_SERVER_SOCKET", req.url);
        });
        socket.on("error", function (err) {
            filterSocketConnReset(err, "CLIENT_TO_PROXY_SOCKET", req.url);
        });
        // Since node 0.9.9, ECONNRESET on sockets are no longer hidden
        // eslint-disable-next-line jsdoc/require-jsdoc
        function filterSocketConnReset(err, socketDescription, url) {
            if (err.code === "ECONNRESET") {
                self._debug.log("Got ECONNRESET on " + socketDescription + ", ignoring. Target: " + url);
            }
            else {
                self._debug.log("Got unexpected error on " + socketDescription + ". Target: " + url, err);
            }
        }
    }
    filterAndCanonizeHeaders(originalHeaders) {
        const headers = {};
        for (const key in originalHeaders) {
            if (originalHeaders.hasOwnProperty(key)) {
                const canonizedKey = key.trim();
                if (/^public\-key\-pins/i.test(canonizedKey)) {
                    // HPKP header => filter
                    continue;
                }
                if (!nodeCommon._checkInvalidHeaderChar(originalHeaders[key])) {
                    headers[canonizedKey] = originalHeaders[key];
                }
            }
        }
        return headers;
    }
    onWebSocketClose(ctx, code, message, callback) {
        // The default behavior of http-mitm-proxy causes exceptions on network errors on websockets
        // so we need to override it
        if (code === 1005 || code === 1006) {
            if (ctx.closedByServer) {
                self._debug.log("ProxyToServer websocket closed due to connectivity issue, terminating ClientToProxy websocket. Target:", ctx.proxyToServerWebSocket.url);
                return ctx.clientToProxyWebSocket.terminate();
            }
            else {
                self._debug.log("ClientToProxy websocket closed due to connectivity issue, terminating ProxyToServer websocket. Target:", ctx.proxyToServerWebSocket.url);
                return ctx.proxyToServerWebSocket.terminate();
            }
        }
        return callback();
    }
};
NtlmProxyMitm = __decorate([
    inversify_1.injectable(),
    __param(0, inversify_1.inject(dependency_injection_types_1.TYPES.IConfigStore)),
    __param(1, inversify_1.inject(dependency_injection_types_1.TYPES.IPortsConfigStore)),
    __param(2, inversify_1.inject(dependency_injection_types_1.TYPES.IConnectionContextManager)),
    __param(3, inversify_1.inject(dependency_injection_types_1.TYPES.NewableIWinSsoFacade)),
    __param(4, inversify_1.inject(dependency_injection_types_1.TYPES.INegotiateManager)),
    __param(5, inversify_1.inject(dependency_injection_types_1.TYPES.INtlmManager)),
    __param(6, inversify_1.inject(dependency_injection_types_1.TYPES.IUpstreamProxyManager)),
    __param(7, inversify_1.inject(dependency_injection_types_1.TYPES.IDebugLogger)),
    __metadata("design:paramtypes", [Object, Object, Object, Object, Object, Object, Object, Object])
], NtlmProxyMitm);
exports.NtlmProxyMitm = NtlmProxyMitm;
